llms/mlx_lm/models/plamo2.py

```diff
@@ -1054,113 +374,53 @@ def __init__(self, config: ModelArgs, layer_idx: int) -> None:
-            past_states.update_attention(key_states_new, value_states_new, self.layer_idx)
-
-        kv_seq_len = key_states.shape[-2]
-        position_ids = mx.arange(kv_seq_len, dtype=mx.int64)[None]
-        q_position_ids = position_ids[:, -query_states.shape[2] :]
-        cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
-        query_states = _rotary_pos_emb(query_states, cos, sin, q_position_ids)
-        key_states = _rotary_pos_emb(key_states, cos, sin, position_ids)
-        # [bsz, nh, t, hd]
-
-        # expand shared kv
-        assert self.k_num_heads == self.v_num_heads
-        key_states = mx.repeat(key_states, self.n_group, 1)
-        value_states = mx.repeat(value_states, self.n_group, 1)
-
-        full_attn = self.layer_idx in self.config.full_attention_idx
-
-        query_states = query_states.astype(attn_dtype)
-        key_states = key_states.astype(attn_dtype)
-        value_states = value_states.astype(attn_dtype)
-        if attention_mask is not None and attention_mask.dtype != bool:
-            attention_mask = attention_mask.astype(attn_dtype)
-        if attention_mask is None:
-            if not full_attn:
-                assert key_states.shape[2] <= self.config.attention_window_size + 1
-            mask = create_attention_mask(hidden_states)
-            attn_output = mx.fast.scaled_dot_product_attention(
-                query_states,
-                key_states,
-                value_states,
-                scale=self.scale,
-                mask=mask,
-            )
-        else:
-            if attention_mask.dtype == bool:
-                attention_mask = mx.where(attention_mask, mx.array(0.0, dtype=mx.float16), float("-inf"))
-            if len(attention_mask.shape) == 2:
-                attention_mask = attention_mask[None, None]
-            assert len(attention_mask.shape) == 4
-
-            if not full_attn:
-                m_swa = swa_mask(
-                    query_states.shape[2],
-                    key_states.shape[2],
-                    self.config.attention_window_size,
-                )
-                # `generate` function creates attention mask that does not consider sliding window
-                m_swa = m_swa[None, None]
-                attention_mask = attention_mask[:, :, -query_states.shape[2] :, -key_states.shape[2] :]
-                attention_mask = mx.where(m_swa, attention_mask, float("-inf"))
-
-            # like AttentionMaskConverter._unmask_unattended in huggingface.transfoermers,
-            # we need to attend to all tokens in masked rows for `scaled_dot_product_attention`
-            bool_mask = mx.logical_not(mx.isneginf(attention_mask))
-            valid_tokens = mx.sum(bool_mask, axis=-1).astype(mx.bool_)  # type: ignore  # (..., q_len)
-            attention_mask = mx.where(valid_tokens[..., None], attention_mask, float(0.0))
```

Comment:
@awni I think we need to revert this part for Sliding Window Attention adopted in PLaMo2?

Reply:
No we should just use the `RotatingKVCache` instead in the `make_cache` method. I left a little TODO there.

Reply:
Ah, that makes sense. Thank you for letting me know that class.